### PR TITLE
fix(nextjs): await headers and cookies as expected in Next.js 15

### LINKED ISSUE #11992: next@canary requires headers to be awaited
### Environment

```
  System:
    OS: Linux 6.8 Ubuntu 22.04.4 LTS 22.04.4 LTS (Jammy Jellyfish)
    CPU: (16) x64 AMD Ryzen 7 5800H with Radeon Graphics
    Memory: 9.95 GB / 14.98 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  npmPackages:
    next: ^15.0.0-canary.179 => 15.0.0-canary.179 
    next-auth: ^5.0.0-beta.22 => 5.0.0-beta.22 
    react: ^19.0.0-rc-1460d67c-20241003 => 19.0.0-rc-1460d67c-20241003 

```

### Reproduction URL

https://github.com/vercel/next.js/blob/canary/examples/auth

### Describe the issue

with recent next@canary, it's expected that certain functions are now awaited, such as headers:
`
In route / a header property was accessed directly with `headers().get('x-forwarded-proto')`. `headers()` should be awaited before using its value. Learn more: https://nextjs.org/docs/messages/sync-dynamic-apis
`

### How to reproduce

1. Clone the next-auth/next example repo
2. update next to canary, react to rc, and next-auth to latest beta. 
3. fill in .env
4. yarn dev

### Expected behavior

no errors or warnings

### PR DESCRIPTION
**Pull request by Code-DJ**
Fixes: #11992

Added await for headers and cookies.

Ran the following script to find issues:
```sh
bunx @next/codemod@canary next-async-request-api .
```

### CODE REVIEW

File: packages/next-auth/src/lib/index.ts [hunk 1]
```diff
@@ -183,7 +183,7 @@ export function initAuth(
   return (...args: WithAuthArgs) => {
     if (!args.length) {
       // React Server Components
-      return getSession(headers(), config).then((r) => r.json())
+      return getSession((headers() as unknown as UnsafeUnwrappedHeaders), config).then((r) => r.json())
```
Comment (balazsorban44) [inline packages/next-auth/src/lib/index.ts:186]: let's chain this with `.then` instead to be able to get rid of the unsafe type

**Review by balazsorban44 — CHANGES_REQUESTED** [review]

**Comment by neoromantic** [comment]
Eagerly waiting for this one to be merged!

**Comment by iFlyinq** [comment]
Thanks for this contribution! ;)

**Comment by iFlyinq** [comment]
If you'd like to say something @richardasymmetric feel free to do so, instead of disliking without feedback :P

**Comment by richardasymmetric** [comment]
Sorry for adding to the noise, everyone subscribed!

Hi @iFlyinq, you're one of [today's lucky 10,000](https://xkcd.com/1053/)! 😊

> If you'd like to say something @richardasymmetric feel free to do so, instead of disliking without feedback :P

I appreciate your enthusiasm, and I see how the comment was meant to express gratitude! However, I tend to thumbs down comments that, while maybe positive, don't necessarily add new technical information to the discussion. This includes comments like "me too," "+1," or "when will this be fixed?". GitHub notifies everyone subscribed to the issue for every comment, and even well-meaning comments like "Thanks for this contribution!" or similar can add to the noise for people who are closely following the thread for updates or technical input.

If you'd like to show support or appreciation for the issue, reactions (👍, 👎, etc.) are a great alternative. They help keep the thread focused while still letting contributors know their work is appreciated. Thanks for understanding, and I hope this helps clarify my reasoning!

File: packages/next-auth/src/lib/actions.ts [hunk 2]
```diff
@@ -71,7 +71,7 @@ export async function signIn(
   const req = new Request(url, { method: "POST", headers, body })
   const res = await Auth(req, { ...config, raw, skipCSRFCheck })
 
-  for (const c of res?.cookies ?? []) cookies().set(c.name, c.value, c.options)
+  for (const c of res?.cookies ?? []) (await cookies()).set(c.name, c.value, c.options)
```
Comment (gnoff) [inline packages/next-auth/src/lib/actions.ts:74]: probably better off pulling this out of the loop

File: packages/next-auth/src/lib/actions.ts [hunk 3]
```diff
@@ -107,7 +107,7 @@ export async function signOut(
 
   const res = await Auth(req, { ...config, raw, skipCSRFCheck })
 
-  for (const c of res?.cookies ?? []) cookies().set(c.name, c.value, c.options)
+  for (const c of res?.cookies ?? []) (await cookies()).set(c.name, c.value, c.options)
```
Comment (gnoff) [inline packages/next-auth/src/lib/actions.ts:110]: same here

**Comment by balazsorban44** [comment]
A fix for this should be out in `next-auth@5.0.0-beta.24` :tada:!

**Pull request closed without merging**